### ISSUE TITLE
another failed test case with BIG numbers in weights and biases

### ISSUE THREAD

**Issue by fleapapa** [01org/daal#16]
I modified the sample code cpp/mpi/sources/neural_net_dense_allgather_distributed_mpi.cpp to train it with a data set with 6 features and 101 classes. When the training finished, prediction test failed on all test samples. The printTensors() call shows "nan" probability on all classes. How come?

I am curious what happened during the training, so i insert another printTensors() in the for-loop on line# 240 (the line has a comment /* FORWARD PASS */) to dump the wTensor[l] and bTensor[l] arrays on each iteration (batch). I am surprised to find that the weights and biases grow larger and larger and eventually became all "NaN".

For example, on the 5th iteration (ie., the one with 100th training sample), the two tensors on layer 0 shows the following values:

wTensor[0] i=100                                                                          bTensor[0] i=100
-0.310         0.504          0.008          0.005          0.027          0.004          0.120
-0.305         0.496          0.008          0.004          0.027          0.002          0.228
-0.304         0.496          0.009          0.004          0.027          0.003          0.228
-0.309         0.503          0.008          0.005          0.028          0.003          0.501
-0.304         0.496          0.009          0.003          0.027          0.003          0.190
-0.306         0.498          0.009          0.005          0.027          0.004          0.293
-0.303         0.492          0.009          0.004          0.026          0.003          0.140
-0.307         0.499          0.008          0.004          0.027          0.004          0.421
-0.314         0.510          0.009          0.003          0.028          0.004          0.173
-0.303         0.495          0.009          0.005          0.025          0.003          0.274
-0.309         0.503          0.009          0.004          0.026          0.003          0.382
-0.306         0.498          0.008          0.005          0.027          0.004          0.025
-0.300         0.488          0.008          0.004          0.025          0.002          0.467
-0.304         0.495          0.008          0.004          0.026          0.003          0.228
-0.315         0.511          0.008          0.005          0.028          0.003          0.462
-0.297         0.484          0.008          0.004          0.027          0.003          0.339

Then, on 10th iteration, the two tensors show the following values:

wTensor[0] i=250                                                                          bTensor[0] i=250
20423804928.000-24076877824.000-2788184064.000-2061500416.000-1226406912.000-257311488.000 -344658944.000
20031678464.000-23614617600.000-2734652672.000-2021921152.000-1202860800.000-252371264.000 -338041728.000
20134170624.000-23735443456.000-2748644608.000-2032265984.000-1209015040.000-253662512.000 -339771296.000
20345231360.000-23984257024.000-2777457920.000-2053569920.000-1221688704.000-256321616.000 -343333024.000
20118687744.000-23717189632.000-2746531072.000-2030703488.000-1208085248.000-253467488.000 -339510048.000
20129929216.000-23730440192.000-2748065024.000-2031837696.000-1208760192.000-253609056.000 -339699648.000
19878868992.000-23434477568.000-2713792000.000-2006497152.000-1193684864.000-250446096.000 -335463008.000
20237498368.000-23857250304.000-2762750976.000-2042696064.000-1215219840.000-254964336.000 -341515040.000
20707330048.000-24411115520.000-2826889728.000-2090118144.000-1243431936.000-260883520.000 -349443520.000
20010489856.000-23589636096.000-2731759872.000-2019782144.000-1201588096.000-252104320.000 -337684096.000
20424769536.000-24078016512.000-2788315904.000-2061597952.000-1226464640.000-257323664.000 -344675232.000
20189376512.000-23800524800.000-2756180992.000-2037838592.000-1212330112.000-254358064.000 -340702912.000
19812022272.000-23355670528.000-2704665856.000-1999749760.000-1189670528.000-249603936.000 -334334880.000
20055734272.000-23642976256.000-2737936384.000-2024348928.000-1204305152.000-252674352.000 -338447648.000
20682610688.000-24381980672.000-2823515392.000-2087623680.000-1241947648.000-260572096.000 -349026432.000
19703820288.000-23228116992.000-2689894400.000-1988828288.000-1183173376.000-248240688.000 -332508960.000
20096104448.000-23690567680.000-2743447808.000-2028423680.000-1206729088.000-253182912.000 -339128864.000                                           
20273483776.000-23899676672.000-2767663104.000-2046327936.000-1217380608.000-255417696.000 -342122272.000
20409819136.000-24060395520.000-2786274560.000-2060088960.000-1225567104.000-257135312.000 -344422944.000
20317394944.000-23951441920.000-2773657600.000-2050760192.000-1220017280.000-255970912.000 -342863264.000

On the next batch, the same layer (0) has the following weights and biases:

wTensor[0] i=275                                                                          bTensor[0] i=275
-14304945344528515072.00013577232474173341696.000262415045722374144.000224360450068840448.0001348394056182398976.00078858193714806784.000234582867370311680.000
-14032691971840016384.00013318827450436681728.000257420617232416768.000220090324863680512.0001322731454790107136.00077357351752957952.000230118248866119680.000
-14100557128041234432.00013383242339149938688.000258665762611265536.000221154823918059520.0001329128413440507904.00077731486354112512.000231231143611990016.000
-14253787267511353344.00013528676941179125760.000261476578188328960.000223558064278601728.0001343571873060880384.00078576168982282240.000233743939998318592.000
-14090812156484255744.00013373992147825459200.000258486920173060096.000221001957442060288.0001328209908914454528.00077677756313239552.000231071388008448000.000
-14101796277645737984.00013384417717080031232.000258688491578195968.000221174288709844992.0001329245236550959104.00077738315352113152.000231251467397234688.000
-13925934890341105664.00013217501956378984448.000255462249584394240.000218415940453269504.0001312668449494794240.00076768838154190848.000228367585836531712.000
-14176242010939195392.00013455075632815800320.000260054067840024576.000222341884159197184.0001336262594637332480.00078148699477245952.000232472337620926464.000
-14501940444830629888.00013764206625499906048.000266028762485751808.000227450129282498560.0001366963158308093952.00079944159015731200.000237813335611932672.000
-14017006338958163968.00013303940062996594688.000257132923143061504.000219844343496704000.0001321252886528655360.00077270885471354880.000229861031864696832.000
-14306688070458540032.00013578887239173144576.000262447051818663936.000224387834780319744.0001348558433170751488.00078867814441549824.000234611471852503040.000
-14139290723664527360.00013420004510424629248.000259376201741631488.000221762286912536576.0001332779616678445056.00077944997768331264.000231866334915330048.000
-13878239175439810560.00013172235062663446528.000254587416285806592.000217667980488605696.0001308172683887771648.00076505926026133504.000227585472291930112.000
-14047856436210302976.00013333220057644269568.000257698845213851648.000220328180152532992.0001324160819906215936.00077440948996407296.000230366927472558080.000
-14490277924994809856.00013753135642919829504.000265814838754672640.000227267198035427328.0001365863921558224896.00079879863355310080.000237622140847783936.000
-13798587254588833792.00013096632643137568768.000253126148152623104.000216418626041806848.0001300664531298549760.00076066799979855872.000226279252478132224.000
-14074563573648982016.00013358569298222645248.000258188849442717696.000220747111262584832.0001326678289216962560.00077588171885379584.000230804911057534976.000
-14199532965750374400.00013477180214580609024.000260481279647023104.000222707093818310656.0001338457907041140736.00078277084639657984.000232854228933017600.000
-14296670420017872896.00013569376463592882176.000262263175678787584.000224230604617547776.0001347614227560398848.00078812572572188672.000234447180763496448.000
-14231540848746561536.00013507560820367687680.000261068436036124672.000223209106775736320.0001341474966947758080.00078453539076046848.000233379108296327168.000

Then, on 12th batch (i=300), all weights and biases become "nan" !!!

wTensor[0] i=300                                                                          bTensor[0] i=300
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan                                                      
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan
-nan           -nan           -nan           -nan           -nan           -nan           -nan

Apparently something is wrong either in the sample code or in DAAL library. I'm trying to figure out in which...but if anyone who is more familiar with the code base could take a look of the issue, i will very appreciate it!

**Comment by fleapapa** [comment]
Is the activation function the root cause? The neural net in current DAAL library seems using ReLU as the activation function. 

Is it possible that the "delta rule" (induced from chain rule) based on ReLU function generates these BIG derivatives (gradient) from my training sample data?...

How can i change the sample code to use Sigmoid or Tanh instead of ReLU? Any pointer will be deeply appreciated.

**Comment by fleapapa** [comment]
After reducing learning rate, the problem is gone.
